paddle/fluid/inference/api/quantizer.cc

```diff
@@ -0,0 +1,433 @@
+    const LoDTensor& var_tensor, bool is_unsigned) const {
+  PADDLE_ENFORCE(var_tensor.dims().size() > 0, "Tensor dimension is empty.");
+
+  ConstEigenVectorArrayMap eigen_tensor{var_tensor.data<float>(),
+                                        var_tensor.numel(), 1};
+  float min_val = eigen_tensor.minCoeff();
+  if (is_unsigned)
+    PADDLE_ENFORCE(
+        min_val >= 0.0f,
+        "Tensor is claimed to be unsigned, but its min value (%f) is < 0.0",
+        min_val);
+
+  int channels = var_tensor.dims()[0];
+  LoDTensor scale_tensor;
+  scale_tensor.Resize({channels});
+  auto* scale_ptr = scale_tensor.mutable_data<double>(CPUPlace());
+
+  for (int i = 0; i < channels; ++i) {
+    const auto tensor = var_tensor.Slice(i, i + 1);
+
+    ConstEigenVectorArrayMap eigen_tensor{tensor.data<float>(), tensor.numel(),
+                                          1};
+    float max_abs = eigen_tensor.abs().maxCoeff();
+    scale_ptr[i] = 1.0 / max_abs;
+  }
+
+  return std::make_pair(is_unsigned, scale_tensor);
+}
+
+std::pair<std::vector<int>, float> AnalysisPredictor::Quantizer::Histogram(
+    const framework::LoDTensor& var_tensor, float min_val, float max_val,
+    size_t num_bins) const {
+  PADDLE_ENFORCE_GT(num_bins, 0,
+                    "Quantizer: To calculate Histogram, num_bins (" +
+                        std::to_string(num_bins) + ") must be positive.");
+  PADDLE_ENFORCE_GT(
+      var_tensor.numel(), 0,
+      "Quantizer: To calculate Histogram, the tensor must not be empty.");
+  PADDLE_ENFORCE(max_val >= min_val,
+                 "Quantizer: To calculate Histogram, max_val (" +
+                     std::to_string(max_val) +
+                     ") must be greater or equal"
+                     "to min_val (" +
+                     std::to_string(min_val) + ").");
+  ConstEigenVectorArrayMap eigen_tensor{var_tensor.data<float>(),
+                                        var_tensor.numel(), 1};
+  auto bin_width = std::abs(max_val - min_val) / num_bins;
+  std::vector<int> hist(num_bins);
+
+  for (int i = 0; i < eigen_tensor.size(); i++) {
+    int bin = std::min(
+        num_bins - 1,
+        static_cast<size_t>(floor((eigen_tensor[i] - min_val) / bin_width)));
+    ++hist[bin];
+  }
+
+  return std::make_pair(std::move(hist), std::move(bin_width));
+}
+
+void AnalysisPredictor::Quantizer::PrepareArgument() const {
+  auto& arg = predictor_.argument_;
+  if (!arg.scope_valid()) arg.SetScope(new framework::Scope);
+  arg.SetMainProgramNotOwned(predictor_.inference_program_.get());
+  auto graph = std::unique_ptr<Graph>(new Graph(arg.main_program()));
+  arg.SetMainGraph(graph.release());
+  arg.main_graph().Set(framework::ir::kParamScopeAttr,
+                       new framework::Scope*(arg.scope_ptr()));
+
+  auto* builder = predictor_.config_.pass_builder();
+  builder->AnalysisPasses().clear();
+  builder->SetPasses({
+      "infer_clean_graph_pass", "cpu_quantize_pass", "cpu_quantize_squash_pass",
```

Comment:
How about using `AppendPass` here?

Reply:
We could, but it would require removing all other passes one by one in a loop first, and then appending also one by one. We can do that if you like.

Reply:
I think you want to insert `cpu_quantize_pass` and `cpu_quantize_squash_pass` but remain all other cpu&mkldnn passes, do I understand right?
Besides, why do you `builder->AnalysisPasses().clear();`? there is no AnalysisPasses here.

Reply:
No, we need only the three passes (`infer_clean_graph_pass`, `cpu_quantize_pass`, `cpu_quantize_squash_pass`) here, as the other passes were applied during the standard optimization phase.

After a call to `pass_builder()` the `analysis_passes_` member of `CpuPassStrategy` contains the default passes (`ir_graph_build_pass`, `ir_analysis_pass`, `ir_params_sync_among_devices_pass`), which are redundant during quantization.